### PR TITLE
Improve typing of autocomplete/choices and docstring parser

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

small typing improvements, cherrypicked from a wip branch

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `pre-commit run --all-files`
- [ ] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Review by EQUENOS — APPROVED** [review]

**Review by Victorsitou — APPROVED** [review]

File: disnake/ext/commands/params.py [hunk 1]
```diff
@@ -61,12 +61,17 @@
...
+    from disnake.app_commands import Choices
+    from disnake.types.interactions import ApplicationCommandOptionChoiceValue
+
     from .slash_core import InvokableSlashCommand, SubCommand
 
     AnySlashCommand = Union[InvokableSlashCommand, SubCommand]
 
     from typing_extensions import TypeGuard
 
+    TChoice = TypeVar("TChoice", bound=ApplicationCommandOptionChoiceValue)
```
Comment (onerandomusername) [inline disnake/ext/commands/params.py:73]: I'm pretty sure this should go outside of the TYPE_CHECKING, and use a string bind, if that's possible.

Comment (shiftinv) [reply disnake/ext/commands/params.py:73]: Why should it? It's not being used at runtime anymore, since the unnecessary `cast` got removed

Comment (onerandomusername) [reply disnake/ext/commands/params.py:73]: Oh wait, yeah, this is fine.

**Review by Victorsitou — COMMENTED** [review]
Wait, I think `ParamInfo.choices` documentation is missing.

image

File: disnake/interactions/base.py [hunk 2]
```diff
@@ -1017,15 +1013,15 @@ async def autocomplete(
...
+            choices_data = []
         elif isinstance(choices, Mapping):
-            data["choices"] = [{"name": n, "value": v} for n, v in choices.items()]
+            choices_data = [{"name": n, "value": v} for n, v in choices.items()]
         elif isinstance(choices, Iterable) and not isinstance(choices[0], OptionChoice):
-            data["choices"] = [{"name": n, "value": n} for n in choices]
+            choices_data = [{"name": n, "value": n} for n in choices]  # type: ignore
         else:
-            data["choices"] = [c.to_dict() for c in choices]  # type: ignore
+            choices_data = [c.to_dict() for c in choices]  # type: ignore
```
Comment (thesadru) [inline disnake/interactions/base.py:1024]: how about doing it after
```py
def _parse_command_choice(choices: Choices) -> t.List[hikari.CommandChoice]:
    """Parse an arbitrary choice into a list of CommandChoice"""
    if isinstance(choices, t.Mapping):
        return [hikari.CommandChoice(name=str(name), value=value) for name, value in choices.items()]

    cc = []
    for value in choices:
        if not isinstance(value, hikari.CommandChoice):
            value = hikari.CommandChoice(name=str(value), value=value)
        cc.append(value)

    return cc
```

File: disnake/ext/commands/params.py [hunk 3]
```diff
@@ -271,7 +273,7 @@ class ParamInfo:
         The name of this slash command option.
     description: :class:`str`
         The description of this slash command option.
-    choices: Iterable[Any]
+    choices: Union[List[:class:`OptionChoice`], List[Union[:class:`str`, :class:`int`]], Dict[:class:`str`, Union[:class:`str`, :class:`int`]]]
```
Comment (Victorsitou) [inline disnake/ext/commands/params.py:276]:
```suggestion
    choices: Union[List[:class:`.OptionChoice`], List[Union[:class:`str`, :class:`int`]], Dict[:class:`str`, Union[:class:`str`, :class:`int`]]]
```

File: disnake/ext/commands/params.py [hunk 4]
```diff
@@ -822,7 +824,7 @@ def Param(
     description: :class:`str`
         The description of the option. You can skip this kwarg and use docstrings. See :ref:`param_syntax`.
         Kwarg aliases: ``desc``.
-    choices: Iterable[Any]
+    choices: Union[List[:class:`OptionChoice`], List[Union[:class:`str`, :class:`int`]], Dict[:class:`str`, Union[:class:`str`, :class:`int`]]]
```
Comment (Victorsitou) [inline disnake/ext/commands/params.py:827]:
```suggestion
    choices: Union[List[:class:`.OptionChoice`], List[Union[:class:`str`, :class:`int`]], Dict[:class:`str`, Union[:class:`str`, :class:`int`]]]
```

**Review by Victorsitou — COMMENTED** [review]
Sorry, I missed this in my last review.
image

File: disnake/interactions/base.py (hunk 2 above)
Comment (shiftinv) [reply disnake/interactions/base.py]: yea, I mostly wanted to get the typing out of the way first, but that part needs a rework.
Should be fixed now

File: disnake/ext/commands/params.py (hunk 4 above)
Comment (shiftinv) [reply disnake/ext/commands/params.py]: fixed, thanks!

**Review by thesadru — APPROVED** [review]

**Pull request merged**